File: hw/pazmino/u3/HW30URIs/HW30URI/src/main/java/ec/edu/espe/hw30uri/view/URIsApp.java
package ec.edu.espe.hw30uri.view;

import ec.edu.espe.hw30uri.controller.ProjectController;
import ec.edu.espe.hw30uri.model.Project;
import org.bson.Document;
import static spark.Spark.*;

/**
 *
 * @author Brandon Pazmiño
 */
public class URIsApp {

    public static void main(String[] args) {
        port(6407);

        get("/infoproject", (req, res) -> {
            String projectId = req.queryParams("projectId");
            if (projectId == null) {
                res.status(400);
                return "{ \"error\": \"Se requiere el parámetro 'projectId'\" }";
            }

            Document project = ProjectController.getProject(projectId);
            if (project == null) {
                res.status(404);
                return "{ \"error\": \"Proyecto no encontrado\" }";
            }

            res.type("application/json");
            return project.toJson();
        });

        post("/addproject", (req, res) -> {
            String projectId = req.queryParams("projectId");
            String title = req.queryParams("title");
            String description = req.queryParams("description");
            String startDate = req.queryParams("startDate");
            String closingDate = req.queryParams("closingDate");
            double startQuote = Double.parseDouble(req.queryParams("startQuote"));
            String operationalStatus = req.queryParams("operationalStatus");
            String quoteStatus = req.queryParams("quoteStatus");
            boolean paid = Boolean.parseBoolean(req.queryParams("paid"));
            boolean invoiced = Boolean.parseBoolean(req.queryParams("invoiced"));
            boolean isPublic = Boolean.parseBoolean(req.queryParams("isPublic"));

            Project project = new Project(projectId, title, description, startDate, closingDate, startQuote,
                    operationalStatus, quoteStatus, paid, invoiced, isPublic);

            boolean success = ProjectController.addProject(project);
            res.type("application/json");
            return success ? "{ \"message\": \"Proyecto agregado exitosamente\" }" : "{ \"error\": \"No se pudo agregar el proyecto\" }";
        });

        put("/updatestateproject", (req, res) -> {
            String projectId = req.queryParams("projectId");
            String newState = req.queryParams("operationalStatus");

            if (projectId == null || newState == null) {
                res.status(400);
                return "{ \"error\": \"Se requieren los parámetros 'projectId' y 'status'\" }";
            }

            boolean success = ProjectController.updateProjectStatus(projectId, newState);
            res.type("application/json");
            return success ? "{ \"message\": \"Estado del proyecto actualizado\" }" : "{ \"error\": \"No se pudo actualizar el estado\" }";
        });

        delete("/deleteproject", (req, res) -> {
            String projectId = req.queryParams("projectId");

            if (projectId == null) {
                res.status(400);
                return "{ \"error\": \"Se requiere el parámetro 'projectId'\" }";
            }

            boolean success = ProjectController.deleteProject(projectId);
            res.type("application/json");
            return success ? "{ \"message\": \"Proyecto eliminado\" }" : "{ \"error\": \"No se pudo eliminar el proyecto\" }";
        });

        System.out.println("Servidor web en http://localhost:6407");
        System.out.println("http://localhost:6407/infoproject?projectId=Prj-0001");
        System.out.println("http://localhost:6407/addproject?projectId=Prj-0020&title=Nuevo%20Proyecto&description=Test...");
        System.out.println("http://localhost:6407/updatestateproject?projectId=Prj-0001&status=En%20progreso");
        System.out.println("http://localhost:6407/deleteproject?projectId=Prj-0001");
    }
}

File: hw/pazmino/u3/HW30URIs/hw40URIPython/ec/espe/espe/hw40_uri_python/view/uris_app.py
from flask import Flask, request, jsonify
from ec.espe.espe.hw40_uri_python.controller.project_controller import ProjectController
from ec.espe.espe.hw40_uri_python.model.project import Project

app = Flask(__name__)

@app.route("/infoproject", methods=["GET"])
def get_project():
    project_id = request.args.get("projectId")
    if not project_id:
        return jsonify({"error": "Se requiere el parámetro 'projectId'"}), 400
    
    project = ProjectController.get_project(project_id)
    if project is None:
        return jsonify({"error": "Proyecto no encontrado"}), 404

    project["_id"] = str(project["_id"])  # Convertir ObjectId a string para evitar problemas con JSON
    return jsonify(project)

@app.route("/addproject", methods=["POST"])
def add_project():
    data = request.form
    
    try:
        project = Project(
            data["projectId"], data["title"], data["description"], 
            data["startDate"], data["closingDate"], float(data["startQuote"]),
            data["operationalStatus"], data["quoteStatus"], 
            data.get("paid", "false").lower() == "true", 
            data.get("invoiced", "false").lower() == "true", 
            data.get("isPublic", "false").lower() == "true"
        )
    except KeyError:
        return jsonify({"error": "Faltan parámetros requeridos"}), 400
    
    success = ProjectController.add_project(project)
    return jsonify({"message": "Proyecto agregado exitosamente"} if success else {"error": "No se pudo agregar el proyecto"})

@app.route("/updatestateproject", methods=["PUT"])
def update_project_state():
    project_id = request.args.get("projectId")
    new_state = request.args.get("operationalStatus")
    
    if not project_id or not new_state:
        return jsonify({"error": "Se requieren los parámetros 'projectId' y 'operationalStatus'"}), 400
    
    success = ProjectController.update_project_status(project_id, new_state)
    return jsonify({"message": "Estado del proyecto actualizado"} if success else {"error": "No se pudo actualizar el estado"})

@app.route("/deleteproject", methods=["DELETE"])
def delete_project():
    project_id = request.args.get("projectId")
    
    if not project_id:
        return jsonify({"error": "Se requiere el parámetro 'projectId'"}), 400
    
    success = ProjectController.delete_project(project_id)
    return jsonify({"message": "Proyecto eliminado"} if success else {"error": "No se pudo eliminar el proyecto"})

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=6407, debug=True)
